### ISSUE TITLE
Missing ARIA markup in the desktop megamenu

### ISSUE THREAD

**Issue by laflannery** [department-of-veterans-affairs/va.gov-cms#10698]
**BLOCKED:** for determination on defect level / urgency level for next steps, Laura Flannery to approve.

## Describe the defect
The desktop megamenu is designed as a treeview navigation, however it is missing the necessary `aria` markup in order to make this navigation fully accessible to all users, including both keyboard and screen reader users.

## Additional Context
A treeview widget provides a hierarchical list of items that can be expanded or collapsed.  It is also important when using this type of navigation, that the focus is managed correctly when an item within the tree is activated - when an item is activated, the focus should move to the beginning of the new content.

The current megamenu is not utilizing the proper roles for this treeview widget and therefore it is not clear to screen readers users when new content has appeared on the right side of the megamenu after a link has been activated.

## Expected behavior

When I activate the top-level "VA Benefits and Health Care" menu option:
- Screen reader indicates/announces that the tree item has been expanded
- Focus moves to the first item in the newly opened content
- When I press the up/down arrows, focus travels through the tree level I am currently on
- When I press the left/right arrow keys, the tree item opens/closes and I move to the next/previous tree level
- **Example to make this arrow navigation clearer:**
  - I first item I tab to in the navigation is "VA Benefits and Health Care".  Using the right arrow key, I expand the sub-nav which displays all the benefits hubs links, with focus on "Health care" (the first item).  I use the down arrow key to navigate to "Careers and Employment". Once there, I use the right arrow key again to expand that section.
- To select a single link, not an expandable tree section, I would use Enter or Space (which are the standard keyboard actions for a link).

When I activate any Hub link (e.g., Disability):
- Focus moves to the sub-menu link, i.e., "View all in [hub name]"
- Screen reader announces the name of the link, i.e. "View all in [hub name]"

### ARIA requirements
- `#vetnav-menu` element should have `role="tree"`
- Each `<li>` with the sub-nav should have `role="none"`
- Each `<button>` (VA Benefits and Health Care & the Hubs) will need to be changed to `<a>` (the treeitem role is not allowed on semantic buttons)
  - Once updated, these new `<a>` tags should have `role="treeitem"`
  - The `aria-controls` attribute should be changed to `aria-owns`
- The sub-navigation containing the links for all the hubs should have `role="group"`
![image](https://user-images.githubusercontent.com/106678594/191041173-1e64fd1b-42ad-4fb6-a1f1-511b82efdc08.png)

## Acceptance Criteria
- [ ] Markup of the top level `<ul>` with `role="tree"`
- [ ] Markup of the first (Hub) level of the megamenu includes aria for treeview widget (see above ARIA requirements)
- [ ] Screen reader interactions with megamenu is as expected (above)
- [ ] Keyboard navigation of megamenu using arrow keys is as described above
- [ ] Confirm with Accessibility Lead 

## Screenshots
Desktop megamenu:
![image](https://user-images.githubusercontent.com/106678594/189696551-b8a8399d-74d6-4f78-a28f-01bf7205cb7a.png)

 ## Resources
[Article ](https://www.w3.org/WAI/ARIA/apg/example-index/treeview/treeview-navigation)on `aria` practices for treeview navigation

**Comment by wesrowe** [comment]
Moving my Q&A with Laura from main body to here to get it out of the way:

### Questions for Laura:
- What is the top level of the tree, the clickable "VA Benefits and Health Care" in the header? And does that get the "role=tree" attribute?
  - The tree should start with the entire navigation `<ul id="vetnav-menu">`, that element should get the tree role
![image](https://user-images.githubusercontent.com/106678594/191036518-d7c2de72-507f-4ccd-8332-f40b7c21cb4f.png)

- Do you know if we need to use computed or declared properties (per docs), or will the default implementation be sufficient? (or am I just confused)
  - I think this question referred to these two links here: 
![image](https://user-images.githubusercontent.com/106678594/191037111-eed98c15-94b2-426c-89a2-90176088e75b.png)
   If that's the case these are for File directories, not navigations like we have so these particular examples do not apply to us.

- Please have a look at the expected behaviors below – the more precise you can be about what is expected, the more efficiently we can develop the right outcome
  -  I have reviewed and updated the ACs.  I also added a code section for update that should happen
- Should we implement this _after_ removing the default-to-Healthcare, so that no sub-menu is open initially? If so, we can state that assumption and write this issue accordingly (e.g., state what buttons should open the submenu of a given Hub)
  - Yes, it would probably be simpler if we assume that the default behavior will be removed.  My comments have assumed this

**Comment by apisandipas** [comment]
After implementing the `aria` mark up as indicated in the ticket, things weren't working. I confirmed with the W3C website that I'd marked things up correctly and then realized it is likely that because of the way we've engineered the mega menu with React, this `treeview` isn't going to work as requested.

Explanation: when a submenu it "hidden" in the mega menu, it doesnt actually exist in the DOM markup until a click event occurs. At which point the subtree in inserted in the DOM. The `aria treeview` paradigm doesn't rely on click events, but rather arrow key presses to expand/collapse it's children. 

My recommendation to achieve `treeview` compatibility would be to change the React implementation to only hide visibly the submenu sections through traditional CSS methods rather than toggle their existence on click events. This would provide a compatible, always-there tree structure in the DOM and not rely on click events for the subtree/submenu items to exist in the DOM. An alternative, but bigger LOE solution to consider at this time would be to replace the Mega Menu entirely with a Web Component-based implementation. Neither of these would be trivial LOE-wise, but the React solution would be lighter-lift in the short-term.

**Comment by jilladams** [comment]
Marking blocked based on conversation with Laura - she'll add more specific notes from accessibility perspective tomorrow morning.

**Comment by laflannery** [comment]
This is not the only issue we have with the header/mega menu, we've had a some others come up, with the injected header, with mobile being separate, etc. As this is not going to be an easy/simple solution I think it may be a better use of time and effort if we put this aside for now in favor of a more holistic approach to the header later on. I want to make sure we are using our efforts the best possible way we can.

**Comment by jilladams** [comment]
From sync with @davidconlon : a 508 defect 1 in Platform terms means: immediately should be resolved. Also meaning: it becomes our top priority and cannot be paused for a bigger/holistic review. In this case, this code / behavior is old / hasn't been touched, and was 508-reviewed before it shipped, so the concern here is the miss in shipping suddenly becoming a fire drill. So we (Laura + Public Websites) should feel empowered to push back on the platform definition of 508 defect 1 for cases like this, where the megamenu. @laflannery will take that up with the Platform accessibility leads tomorrow.

If the level 1 defect isn't negotiable, we will need to urgently prioritize this fix, taking on: rebuilding the menu to hide rather than remove DOM elements, in order to use treeview. 

Sideline, PW will take on a SPIKE around the bigger picture of header improvements needed, including performance / React usage, mobile, and 508 concerns to figure out what is the scope of a holistic approach.

**Comment by laflannery** [comment]
Notes from my a11y leads call today:

## General
After a product or feature launches, there are instances when new issues may be found.  Other teams do run maintenance audits of their products and they may find something related to our team or it may just be something they happen to come across.  Regardless of how it's found, the reason why it might be surfacing now and was not found during the initial Platform Collab Cycle prior to launch could be for a few reasons:

- Updated guidelines or standards
  - The original audit may have been done when the VA standard was WCAG 2.0 but we are now using WCAG 2.1 to audit all products so new issues related to the new criteria are going to be surfaced.
- Updated technology (i.e. browser, screen reader, etc.)
  - Newer versions of browsers, screen readers and combinations of these technologies are more nuanced and can alert us to different issues
- Updated testing processes
  - Previously, Platform did not have a very clear, repeatable testing process.  This is something that has been updated but is still being worked on to try and make it even more standardized.  The ultimate goal is to make this as bullet proof as possible so that 1 tester on 1 day gets the exact same results as another tester on another day but this was not always the case.

When new issues are found on existing code, we do still want to keep the objectivity of the severity rubric, meaning, a color contrast issue is going to be a defect-1 regardless of if it's found pre-launch or post-launch. However, the timeline to resolve that post-launch issue is going to be negotiable, especially if that post-launch issue is:
- Found on a long-standing product/feature that has not been touched since launch
- Complex, nuanced and possibly involves other issues
When a post-launch complex issue is found, that is a defect-1, or even a defect-2 if these issues would be better served by taking a little more time to examine them from a bigger picture lens and creating a roadmap to remediation that may take longer to complete, that is acceptable.

### What this means for this issue specifically
This is a complex, larger issue that was found on code that had not been touched since launch. It is negotiable as far as the timeline to resolve. I do feel, and so did Martha and the other leads on the call, that holding on the "quick" fix and determining a solution that would best serve not only our Veterans but also our team in terms of effort and maintenance would be the best path forward on this particular issue.

**Comment by artsymartha68** [comment]
cc @mnorthuis per request in case this involves IA changes.

**Comment by jilladams** [comment]
@laflannery that all makes sense. Are you open to us starting to use `508-pre-launch` vs. `508-post-launch` to differentiate these `defect-1` tickets for now? Or do you have an alternate label you'd prefer?

**Comment by jilladams** [comment]
Thanks! Also noting: we've added this to an epic to begin aggregating issues for that header iteration: https://github.com/department-of-veterans-affairs/va.gov-cms/issues/9136

**Comment by wesrowe** [comment]
Moving back to Needs refining for the team to address when the time comes

**Comment by laflannery** [comment]
I've cleaned up and reworked the existing header/mega menu markup and it has been implemented in the Federal Header codepen prototype. This will be tested during Phase 2 research which may result in additional insights and updates but otherwise, this new markup will be implemented on the VA.gov site.
